### PR TITLE
DSE 1038 naive namespace checking

### PR DESCRIPTION
**Pull request by dataswifty**
AC:
Context:
Namespace checking in Hat/ADJ allowed for any valid namespace to be written to.This is due to naive case class checking.

Goal:
Allow and disallow the correct operations for a matching namespace.

Solution:
Extract and improve the namespace permission testing and write tests to prove it's correctness.

Environment:
All

## Source
dswift.atlassian.net/browse/DSE-1038

## Type
- [x] Bug Fix
- [ ] Feature Addition 
- [x] Refactor
- [x] HotFix

## Checklist
- [x] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [x] added a test

### CODE REVIEW

File: hat/test/org/hatdex/hat/utils/NamespaceTestSpec.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      val namespace              = "incorrectnamespace"
+      val applicationPermissions = List(NamespaceWrite("correctnamespace"), NamespaceRead("correctnamespace"))
+
+      val readRoles  = NamespaceUtils.testReadNamespacePermissions(applicationPermissions, namespace)
+      val writeRoles = NamespaceUtils.testWriteNamespacePermissions(applicationPermissions, namespace)
+
+      readRoles == false
+      writeRoles == false
+    }
+  }
```
Comment (benshaw) [inline hat/test/org/hatdex/hat/utils/NamespaceTestSpec.scala:37]: Non Blocking: Id like to see more tests in particular what happends when the list is empty

File: hat/app/org/hatdex/hat/utils/NamespaceUtils.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  def testReadNamespacePermissions(
+      roles: Seq[UserRole],
+      namespace: String): Boolean = {
+    val matchedRoles = roles.map {
+      case NamespaceRead(n) if n == namespace => Some(namespace)
+      case _                                  => None
+    }
+    !matchedRoles.flatten.isEmpty
```
Comment (benshaw) [inline hat/app/org/hatdex/hat/utils/NamespaceUtils.scala:49]: Non Blocking: Negations are notorious for maintenance issues, this could potentially be simplified to matchedRoles.flatten.nonEmpty

**Review by benshaw — APPROVED** [review]

**Review by augustinas — DISMISSED** [review]
The logic between testRead and testWrite functions feels repetitive, but I'm not able to offer a better solution quickly. Approving.

File: hat/test/org/hatdex/hat/utils/NamespaceTestSpec.scala (hunk 1 above)
Comment (dataswifty) [reply hat/test/org/hatdex/hat/utils/NamespaceTestSpec.scala:37]: Agreed, I will add them.

File: hat/app/org/hatdex/hat/utils/NamespaceUtils.scala (hunk 2 above)
Comment (augustinas) [reply hat/app/org/hatdex/hat/utils/NamespaceUtils.scala:49]: good catch, I missed it

Comment (dataswifty) [reply hat/app/org/hatdex/hat/utils/NamespaceUtils.scala:49]: Agreed, I'll update.

**Review by benshaw — APPROVED** [review]

**Comment by dataswifty** [comment]
> The logic between testRead and testWrite functions feels repetitive, but I'm not able to offer a better solution quickly. Approving.

Agreed, it's repetitive but clear IMHO.
